hal/src/thumbv7em/clock/v2/osculp32k.rs

```diff
@@ -1,129 +1,125 @@
     }
 
     #[inline]
-    pub(super) fn activate_1k(&mut self, enabled: bool) {
+    fn enable_1k(&mut self, enabled: bool) {
         self.osculp32k().modify(|_, w| w.en1k().bit(enabled));
     }
 
     #[inline]
-    pub(super) fn activate_32k(&mut self, enabled: bool) {
+    fn enable_32k(&mut self, enabled: bool) {
         self.osculp32k().modify(|_, w| w.en32k().bit(enabled));
     }
 
-    #[allow(dead_code)]
     #[inline]
     fn wrtlock(&mut self) {
         self.osculp32k().modify(|_, w| w.wrtlock().bit(true));
     }
 }
 
 //==============================================================================
-// OscUlp32k
+// OscUlpBase
 //==============================================================================
 
-/// Struct representing a disabled ultra low power oscillator
-///
-/// In reality, this oscillator is always enabled and
-/// [`Enabled`]`<`[`OscUlp32k`]`>` does not have a `disable` method that would
-/// allow to retreive underlying [`OscUlp32k`] struct.
-///
-/// User can only activate/deactivate outgoing signals (as in making them
-/// visible outside of an oscillator).
-/// Therefore [`Enabled`]`<`[`OscUlp32k`]`<`[`Inactive32k`]`, `[`Inactive1k`]`>,
-/// _>` represents an **always** enabled oscillator with disabled outgoing
-/// signals.
-///
-/// It is generic over:
-/// - An output state of a 32 kHz signal (active/inactive)
-/// - An output state of a 1 kHz signal (active/inactive)
-pub struct OscUlp32k {
-    pub(super) token: OscUlp32kToken,
+pub struct OscUlpBase {
+    token: OscUlpBaseToken,
 }
 
-impl OscUlp32k {
-    /// Create a new instance of [`OscUlp32k`]
+impl OscUlpBase {
+    /// Create the ultra-low power base oscillator
+    ///
+    /// Safety: There must never be more than one instance of this struct at any
+    /// given time.
     #[inline]
-    pub(crate) fn new(token: OscUlp32kToken) -> Self {
-        Self { token }
+    pub(super) unsafe fn new() -> Enabled<Self, U0> {
+        let token = OscUlpBaseToken(());
+        Enabled::new(Self { token })
     }
 }
 
-impl Enabled<OscUlp32k, U0> {
-    /// Set calibration parameters
-    ///
-    /// This data is used to compensate for process variations
+impl<N: Counter> Enabled<OscUlpBase, N> {
+    /// Override the factory-default calibration value
+    #[inline]
+    pub fn set_calibration(&mut self, calib: u8) {
+        self.0.token.set_calibration(calib);
+    }
+
+    /// Set the write-lock, which will last until POR
     ///
-    /// This method performs a HW register write. At startup, this register is
-    /// populated with a default parametrization from Flash Calibration at
-    /// startup
+    /// This function sets the write-lock bit, which lasts until power-on reset.
+    /// It also consumes and drops the [`XoscBase`], which destroys API access
+    /// to the registers.
     #[inline]
-    pub fn set_calibration(mut self, calib: u8) -> Self {
-        self.0.token.set_calib(calib);
-        self
+    pub fn write_lock(mut self) {
```

Comment:
The write lock *is* POR based in this case. From the documentation:

> It is also possible to lock the XOSC32K configuration by setting the Write Lock bit in the 32 kHz
External Crystal Oscillator Control register (XOSC32K.WRTLOCK=1). If set, the XOSC32K
configuration is locked until a Power-On Reset (POR) is detected.

Is that not your question?

Reply:
I think what @vccggorski means is that this lock is hard to model for any reset/change of application where the lock state is not dynamically picked up.

This might work just fine in a future Dyn-clocktree setup but as soon as the “Locked<X>” type is gone there is a disconnect between the abstraction and hardware.

Example:
Bootloader sets up a minimal clock tree and locks a few clocks here and there. Bootloader hands off to main application which is unaware of the exact clock tree setup. This would be fine if main application never had to touch the clock tree.

But bootloaders in contrast to main applications tends to be small and simple (with matching clocking needs) and thus the odds that modifications to clocking is required is quite high.

Trying to reset and reconfigure the locked parts of clock tree would fail and in the case it just _ignores_ the new setup (not sure about the implications of writing to write locked registers) there would be a huge disconnect between clock abstraction and actual hardware.

Maybe the simplest for now is to keep the method available but `unsafe` and let any users ensure themselves that it works for their specific needs?

Reply:
I think the bootloader problem is a separate issue. Right now, we only provide an API to create a representation of the clock tree at POR. If that representation is not accurate for any reason, then all bets are off. It's essentially UB. Without knowing the state of the clock tree, we can't be sure that *anything* will work correctly.

The solution is to provide an `unsafe` way for users to create their own clock tree initialization, without writing any registers. I think that is what initially led @vccggorski to move all hardware writes to the `.enable()` functions.

Alternatively, if the state of the clock tree cannot be known until run-time, then our existing, compile-time API is insufficient. The only choice is to wait for a future `Dyn` API. I don't see any way around that.

Maybe you're suggesting that parts of the clock tree could be known statically while other parts could only be known dynamically? Even if that's true, you could still create an accurate representation of the static parts.

However, if you can be sure that your clock tree is accurate (which I claim **must** be true, or you essentially have UB), then I think the `write_lock` function accurately models the write lock.